sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/BeamSqlCorrelVariableExpression.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.impl.interpreter.operator;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.values.Row;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+/** A primitive operation for deferencing a correlation variable. */
+public class BeamSqlCorrelVariableExpression extends BeamSqlExpression {
+
+  private final int correlationId;
+
+  public BeamSqlCorrelVariableExpression(SqlTypeName sqlTypeName, int correlationId) {
+    super(null, sqlTypeName);
+    this.correlationId = correlationId;
+  }
+
+  @Override
+  public boolean accept() {
+    return true;
+  }
+
+  @Override
+  public BeamSqlPrimitive evaluate(
+      Row inputRow, BoundedWindow window, ImmutableMap<Integer, Object> correlateEnv) {
```

Comment:
where does correlateEnv get populated?

Reply:
Only in BeamUnnestRel. Other than that, it is just passed around. The meaning is that the main PCollection input is an outer loop, and each time through the current row (or whatever) gets set to the $cor0 in the correlateEnv. Then the inner loop runs the project expression and pulls out the fields from the current row.

Reply:
I see. Would it work for complex cases, like multiple nested correlated queries? What the plan would look like? I don't see where this is handled, maybe I missed it.

If not, then it probably makes sense to comment on correlation logic that it's only there to support unnest and is limited. And is there a need for a map in this case? Wouldn't a single id work?

And should we wrap the map into some kind of execution context class?

Reply:
Oh, I think that the correlation expression - this class - will work in general. It can be nested arbitrarily AFAIK. The place where that support is missing is at the `Rel` level. If we had a general `BeamCorrelateRel` then it would have to add its var to the environment just like the `BeamUnnestRel` does. And to be general, all of the `Rel` bits would also need to also know about the environment and do something clever. I'm not sure that would be the overall solution - likely we want to successfully use Calcite's decorrelate. It did not seem to actually decorrelate this particular query, though I might have just been using it wrong.